File: frontend/src/types/invoice.ts
export type InvoiceStatus = 'CLEAN' | 'FLAGGED' | 'REVIEWED' | string

export type LineItem = {
  id: string
  item_index: number
  description: string
  quantity: number | string
  unit_price: number | string
  line_total: number | string
}

export type InvoiceResponse = {
  id: string
  invoice_number: string
  vendor_name: string
  date: string
  due_date: string | null
  currency: string
  subtotal: number | string
  tax_rate: number | string
  tax_amount: number | string
  grand_total: number | string
  status: InvoiceStatus
  audit_report: string | null
  uploaded_by: string
  processed_at: string | null
  created_at: string
  line_items: LineItem[]
}

export type ValidationCheck = {
  check_name: string
  expected_value: number | string
  actual_value: number | string
  passed: boolean
  discrepancy: number | string
}

export type ValidationResult = {
  all_passed: boolean
  checks: ValidationCheck[]
  audit_report: string | null
}

export type InvoiceUploadResponse = {
  status: InvoiceStatus
  invoice: InvoiceResponse
  validation: ValidationResult
  message: string
}

export type InvoiceListItem = {
  id: string
  invoice_number: string
  vendor_name: string
  date: string
  grand_total: number | string
  currency: string
  status: InvoiceStatus
  processed_at: string | null
}

export type InvoiceListResponse = {
  total: number
  page: number
  per_page: number
  total_pages: number
  invoices: InvoiceListItem[]
}

export type InvoiceListParams = {
  status?: string
  search?: string
  page?: number
  perPage?: number
}

File: backend/schemas/invoice.py
from pydantic import BaseModel, ConfigDict
from typing import List, Optional
from uuid import UUID
from datetime import datetime, date
from decimal import Decimal
from models.invoices import InvoiceStatus
from schemas.validation import ValidationResult 

# Line Item schemas
class LineItemBase(BaseModel):
    item_index: int
    description: str
    quantity: Decimal
    unit_price: Decimal
    line_total: Decimal

class LineItemCreate(LineItemBase):
    invoice_id: UUID

class LineItemResponse(LineItemBase):
    id: UUID
    
    model_config = ConfigDict(
        from_attributes=True
    )



# Invoice schemas
class InvoiceBase(BaseModel):
    invoice_number: str
    vendor_name: str
    date: date
    due_date: date | None = None
    currency: str = "INR"
    subtotal: Decimal
    tax_rate: Decimal
    tax_amount: Decimal
    grand_total: Decimal

class InvoiceCreate(InvoiceBase):
    line_items: List[LineItemCreate]

class InvoiceResponse(InvoiceBase):
    id: UUID
    status: InvoiceStatus
    audit_report: str | None = None
    uploaded_by: UUID
    processed_at: datetime | None = None
    created_at: datetime
    line_items: List[LineItemResponse]
    ai_processed: bool
    ai_message: str | None = None
    
    model_config = ConfigDict(
        from_attributes=True
    )

class InvoiceListItem(BaseModel):
    id: UUID
    invoice_number: str
    vendor_name: str
    date: date
    grand_total: Decimal
    currency: str
    status: InvoiceStatus
    processed_at: datetime | None = None
    ai_processed: bool
    ai_message: str | None = None
    
    model_config = ConfigDict(
        from_attributes=True
    )

class InvoiceListResponse(BaseModel):
    total: int
    page: int
    per_page: int
    total_pages: int
    invoices: List[InvoiceListItem]
 

class InvoiceUploadResponse(BaseModel):
    status: InvoiceStatus
    invoice: InvoiceResponse
    validation: ValidationResult
    message: str


